File: tcp-server/tcp-server.ts
import net from "node:net"

const PORT = 8080
const server = net.createServer()

server.on('connection', (socket: net.Socket) => {
    socket.on('data', (data: Buffer) => {
        console.log(`Data received from client: ${data}`);
        socket.write(`[*] Server echo: ${data}`);
    })

    socket.on('close', () => {
        console.log('[*] Client disconnected')
    })

    socket.on('error', (error: Error) => {
        console.error(`[*] Connection error: ${error}`)
    })
})

server.listen(PORT, () => {
    console.log('[*] Server listening on', server.address())
})

File: tcp-server/tcp-server.py
import socket
import threading

HOST = "127.0.0.1"
PORT = 8080


def main():
    try:
        server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        server.bind((HOST, PORT))
        server.listen()
        print(f"[*] Server listening on {HOST}:{PORT}")
    except socket.error as err:
        print(f"Failed to start server: {err}")
        return

    while True:
        try:
            connection, address = server.accept()
        except socket.error as err:
            print(f"Error accepting connection: {err}")
            continue

        print(f"[*] Connection accepted from {address}")
        connection_handler = threading.Thread(
            target=handle_connection, args=(connection,)
        )
        connection_handler.start()


def handle_connection(connection: socket.socket):
    text = connection.recv(1024)
    message = text.decode("utf8").strip()
    print(f"[*] Received: {message}")
    connection.close()


if __name__ == "__main__":
    main()
